File: MinSpantree.m
//
//  main.m
//  MinSpantree
//
//  Created by Carl Friess on 10/05/2017.
//  Copyright © 2017 Carl Friess. All rights reserved.
//

#import <Foundation/Foundation.h>


@interface Scanner : NSObject
{
    NSFileHandle* source;
    NSScanner* buffer;
}

- (int)nextInt;

@end

@implementation Scanner

- (instancetype)init
{
    self = [super init];
    if (self) {
        source = [NSFileHandle fileHandleWithStandardInput];
    }
    return self;
}

- (int)nextInt
{
    if (buffer == nil || buffer.atEnd) {
        NSString* nextInput = [[NSString alloc] initWithData:source.availableData encoding:NSUTF8StringEncoding];
        if (nextInput != nil) {
            buffer = [[NSScanner alloc] initWithString: nextInput];
        }
    }
    int value = -1;
    [buffer scanInt:&value];
    return value;
}

@end


@interface Vertex : NSObject
{
    int value;
    Vertex* parent;
}

@property Vertex* parent;

- (Vertex*)getRoot;

@end

@implementation Vertex

- (instancetype)initWithValue:(int)defaultValue
{
    self = [super init];
    if (self) {
        value = defaultValue;
        parent = self;
    }
    return self;
}

@synthesize parent;

-(Vertex *)getRoot
{
    
    return parent == self ? self : [parent getRoot];
    
}

@end


@interface Edge : NSObject
{
    Vertex* u;
    Vertex* v;
    int weight;
}

@property Vertex* u;
@property Vertex* v;
@property int weight;

@end

@implementation Edge

@synthesize u;
@synthesize v;
@synthesize weight;

@end


int main(int argc, const char * argv[]) {
    @autoreleasepool {
        // insert code here...
        
        Scanner* scanner = [[Scanner alloc] init];
        
        for (int numTests = [scanner nextInt]; numTests > 0; numTests--) {
            
            int numVertices = [scanner nextInt];
            int numEdges = [scanner nextInt];
            
            NSMutableArray* vertices = [NSMutableArray arrayWithCapacity:numVertices];
            NSMutableArray* edges = [NSMutableArray arrayWithCapacity:numEdges];
            
            for (int i = 0; i < numVertices; i++) {
                [vertices addObject:[[Vertex alloc]initWithValue:i]];
            }
            
            for (int i = 0; i < numEdges; i++) {
                
                int uValue = [scanner nextInt] - 1;
                int vValue = [scanner nextInt] - 1;
                
                Edge* edge = [[Edge alloc] init];
                
                edge.u = vertices[uValue];
                edge.v = vertices[vValue];
                edge.weight = [scanner nextInt];
                
                [edges addObject:edge];
                
            }
            
            [edges sortUsingComparator:^NSComparisonResult(Edge* first, Edge* second) {
                return first.weight > second.weight;
            }];
            
            int totalWeight = 0;
            
            Edge* edge;
            for (edge in edges) {
                if ([edge.u getRoot] != [edge.v getRoot]) {
                    [edge.v getRoot].parent = [edge.u getRoot];
                    totalWeight += edge.weight;
                }
            }
            
            printf("%d\n", totalWeight);
            
        }
        
    }
    return 0;
}

File: MinSpantree.swift
//
//  main.swift
//  MinSpantreeSwift
//
//  Created by Carl Friess on 03/05/16.
//  Copyright © 2016 Carl Friess. All rights reserved.
//

import Foundation

class Scanner {
    let source = NSFileHandle.fileHandleWithStandardInput()
    var buffer: NSScanner?
    func nextInt() -> Int {
        if buffer == nil || buffer!.atEnd {
            if let nextInput = NSString(data: source.availableData, encoding: NSUTF8StringEncoding) {
                buffer = NSScanner(string: nextInput as String)
            }
        }
        var value: Int = -1;
        buffer!.scanInteger(&value)
        return value
    }
}

class Vertex {
    var value: Int = 0
    var parent: Vertex? = nil
    
    init(value: Int) {
        self.value = value
        parent = self
    }
    
    func getRoot() -> Vertex {
        return self.parent === self ? self : self.parent!.getRoot()
    }
    
}

class Edge {
    var u: Vertex? = nil
    var v: Vertex? = nil
    var weight = 0
    
    init(u: Vertex, v: Vertex, weight: Int) {
        self.u = u
        self.v = v
        self.weight = weight
    }
    
}

let scanner = Scanner()
let numTests: Int = scanner.nextInt()

for test in 0..<numTests {
    
    var numVertices = scanner.nextInt()
    var numEdges = scanner.nextInt()
    
    var vertices = [Vertex?](count: numVertices, repeatedValue: nil)
    var edges = [Edge]()
    
    for i in 0..<numEdges {
        
        let uValue = scanner.nextInt() - 1
        let vValue = scanner.nextInt() - 1
        
        if vertices[uValue] == nil {
            vertices[uValue] = Vertex(value: uValue)
        }
        
        if vertices[vValue] == nil {
            vertices[vValue] = Vertex(value: vValue)
        }
        
        edges.append(Edge(u: vertices[uValue]!, v: vertices[vValue]!, weight: scanner.nextInt()))
        
    }
    
    edges.sortInPlace({ $0.weight < $1.weight })
    
    var totalWeight = 0
    
    for edge in edges {
        if edge.u!.getRoot() !== edge.v!.getRoot() {
            edge.v!.getRoot().parent = edge.u!.getRoot()
            totalWeight += edge.weight
        }
    }
    
    print(totalWeight)
    
}
